### PR TITLE
update the storage selfhealing spec

### PR DESCRIPTION
**Pull request by GlenDC**
since it was created, it got a bit out of sync with
the actual planned work and issues in 0-Disk.

this patch aims to update the storage selfhealing spec
such that the work that is planned is reflected in the spec as it is.

### CODE REVIEW

File: specs/selfhealing/storage.md [hunk 1]
```diff
@@ -21,32 +21,47 @@ This will prevent any creation or deletion of new vdisks.
 ### Step 2: notify all NBD servers that a specific storage engine is not functional
 Update all impacted vdisks that a certain shard is repairing.
 
-By means of updating the AYS model (storage_engine service) and the corresponding configuration in the etcd cluster, all NDB servers that are runnning will be notified that the storage engine is faulty (```status: offline```), and that it needs to fail over to the slave cluster for the specific shard that is offline.
+By means of updating the AYS model (storage_engine service) and the corresponding configuration in the etcd cluster, all NDB servers that are runnning will be notified that the storage engine is faulty (```status: offline```), and that it needs to fail over to the slave cluster for the specific shard that is offline. The vdisks which reported this issue (if it existed) will already have marked the relevant shards as `offline`, and thus no noger have to do anything.
```
Comment (FastGeert) [inline specs/selfhealing/storage.md]: Do you mean that the gonbdserver will update the model in etcd as well, or just internally?

File: specs/selfhealing/storage.md [hunk 2]
```diff
@@ -21,32 +21,47 @@ This will prevent any creation or deletion of new vdisks.
...
-If its temporary, as soon as AYS is able get the storage_engine running again (eg, wait until the node has completed rebooting, or restarted the ARDB server after a crash) it will update the AYS model and corresponding etcd configuration and assign a status (```status: repair-tmp-outage```). This will make the running NDB servers start repairing. 
+If its temporary, as soon as AYS is able get the storage_engine running again (eg, wait until the node has completed rebooting, or restarted the ARDB server after a crash) it will update the AYS model and corresponding etcd configuration and assign a status (```status: repair```). This will make the running NDB servers start repairing for the relevant vdisks.
 
-vdisks (NBD servers) that are not running will be repaired by a NBD server started on a node part of the storage network. While this NDB server is repairing, it should not be possible to start a vm on it.
+Vdisks aren't mounted, will have to be repaired using the zeroctl tool suite. For each vdisk that has to be repaired, the `zeroctl copy dataserver` command will have to be called (e.g. ```zeroctl copy dataserver src_ip dst_ip vdiskid```).
```
Comment (FastGeert) [inline specs/selfhealing/storage.md]: Just a language error: Vdisks **that** are not mounted

File: specs/selfhealing/storage.md [hunk 3]
```diff
@@ -21,32 +21,47 @@ This will prevent any creation or deletion of new vdisks.
...
+A vdisk shouldn't be allowed to mount until it has been repaired (in case it had to be repaired unmounted using the zeroctl tool due to not being mounted yet).
 
-When the NDB server has completed repairing the situation, it will emit the following log line to its std-err stream ```{"type": "repair-final-outage-complete", "data": {"address": "10.100.0.1:22005", "vdiskid": "23"}}```, and start to serve blocks again from this storage engine. Via the node.zero-os watchdog action in AYS this problem will be propagated to the vdisk service. If all vdisks completed repairing the storage_engine service will marked permenantly offline (```status: rip```) both in the AYS model and in the etcd cluster. On top of that the storage cluster will be marked healthy again so that vdisk creation and deletion will be operational again.
+Once all vdisks fof the storage_engine have been repaired, the storage_engine service will be marked as online again (```status: online```) both in the AYS model and in the etcd cluster. On top of that the storage cluster will be marked healthy again so that vdisk creation and deletion will be operational again.
```
Comment (FastGeert) [inline specs/selfhealing/storage.md]: I think this is a mistake. The storage_engine is dead. Status **rip** is IMHO still what we need here.

File: specs/selfhealing/storage.md [hunk 4]
```diff
@@ -21,32 +21,47 @@ This will prevent any creation or deletion of new vdisks.
...
+# SSD failure in template Storage Cluster
+When an SSD fails in the template Storage Cluster the NBD has currently no possibility to self-heal from this. It will send an error message to the std-err stream, and make the vdisk that makes use of it fail with an error now and until the template cluster has been fixed.
+
+The message in question it will send is is something like:
+
+```
+2::{"subject":"ardb","status":421,"data":{"address":"1.2.3.4:16379","db":41,"type":"template","vdiskID":"vd2"}}
+```
+
+The status will be any of 421, 422 or 423. See https://github.com/zero-os/0-Disk/blob/master/docs/log.md#status-codes for more information.
```
Comment (FastGeert) [inline specs/selfhealing/storage.md]: This is out of scope for now. The way how we used to deal with templates needs to revised, due to the introduction of the tlog. Before a template can be used, we will need to import it completely. Then local clones will be created of the template. 

To avoid confusion, please remove this section.

File: specs/selfhealing/storage.md [hunk 5]
```diff
@@ -21,32 +21,47 @@ This will prevent any creation or deletion of new vdisks.
...
 
-When the NDB server has completed repairing the situation, it will emit the following log line to its std-err stream ```{"type": "repair-tmp-outage-complete", "data": {"address": "10.100.0.1:22005", "vdiskid": "23"}}```, and start to serve blocks again from this storage engine. Via the node.zero-os watchdog action in AYS this problem will be propagated to the vdisk service. If all vdisks completed repairing the storage_engine service will marked as online again (```status: online```) both in the AYS model and in the etcd cluster. On top of that the storage cluster will be marked healthy again so that vdisk creation and deletion will be operational again.
+A vdisk shouldn't be allowed to mount until it has been repaired (in case it had to be repaired unmounted using the zeroctl tool due to not being mounted yet).
+
+Once all vdisks fof the storage_engine have been repaired, the storage_engine service will be marked as online again (```status: online```) both in the AYS model and in the etcd cluster. On top of that the storage cluster will be marked healthy again so that vdisk creation and deletion will be operational again.
```
Comment (FastGeert) [inline specs/selfhealing/storage.md]: fof ==> of

File: specs/selfhealing/storage.md (hunk 1 above)
Comment (GlenDC) [reply specs/selfhealing/storage.md]: No. I've always said that all 0-Disk code uses the offline etcd cluster as a read-only config source. It will never update it, only read from it and subscribe to it. This also has nothing to do with the gonbdserver.

This is just an internal in-memory update in the model that a vdisk itself has for that cluster.

File: specs/selfhealing/storage.md (hunk 3 above)
Comment (GlenDC) [reply specs/selfhealing/storage.md]: It's a copy-paste mistake. Should be ```status: rip``` indeed. No worries, just a typo :)

**Comment by GlenDC** [comment]
> As a general comment. Only AYS should update the etcd, to prevent race conditions.

Sure, I have never wanted otherwise :) I'm fairly sure I was one of the people saying that since the beginning.

**Comment by GlenDC** [comment]
@FastGeert is there anything you still want to have changed, or is it good as it is now? Just so I know we're on the same page.
